src/main/java/com/google/gcp_variant_transforms/library/VcfParserImpl.java

```diff
@@ -5,17 +5,43 @@
 import com.google.common.collect.ImmutableList;
 import com.google.inject.Singleton;
 import htsjdk.variant.vcf.VCFCodec;
+import htsjdk.variant.vcf.VCFHeader;
 
 /** Implementation of {@link VcfParser} service. */
 @Singleton
 public class VcfParserImpl implements VcfParser {
 
   private static final long serialVersionUID = 6787944179556195968L;
+  private VCFHeader vcfHeader;
+  private VCFCodec vcfCodec;
 
   @Override
   public VCFCodec generateCodecFromHeaderLines(ImmutableList<String> headerLines) {
     VCFCodec vcfCodec = new VCFCodec();
     vcfCodec.readActualHeader(new HeaderIterator(headerLines));
+    setVcfCodec(vcfCodec);
     return vcfCodec;
   }
+
+  public void setVcfCodec(VCFCodec vcfCodec){

```

Comment:
Only one question about the architecture, just in my opinion:  
I think the Impl class is only provided for vcfParser interface, which is a service to be called in the Guice framework(and Guice has binded vcfParser -> Impl)  
So the architecture workflow will be **"Header -> (vcfParser Service) -> VCFCodec -> VCFHeader and VariantContext"**
I do not think we need those getter and setters in Impl class since we are only using the vcfParser Service in Pipeline runner and we never use the Impl class directly, only bind the class to this interface and use interface everywhere. In other word, only those methods implementing the interface with annotation @Override will be used in the Guice framework.  
As a result, to be honest, I have no idea how to generate VCFHeader in the same service level with VCFCodec since VCFCodec always comes first. The way I am using is just calling generatingCodec function to get the codec and then call codec.getHeader() in other classes when VCFHeader is needed.
Only my opinion... I am just find this Impl class methods weird in the whole framework... 
@tneymanov What do you think?

Reply:
So if we don't want to use `AssertInject` approach, which is admittedly more complex, the easier solution is to do the following

add method `public abstract generateVCFHeader(ImmutableList<String> headerLines);` to `VcfParser`.

and add its implementation in `VcfParserImpl`:

```
@Override
public VCFHeader generateVCFHeader(ImmutableList<String> headerLines) {
    VCFCodec vcfCodec = new VCFCodec();
    vcfCodec.readActualHeader(new HeaderIterator(headerLines));
    return vcfCodec.getHeader();
}
```

Then we would modify VcfToBqContext to have getter and setter for VCFHeader (say getVCFHeader/setVCFHeader).

Finally somewhere in the beginning of the Task code after reading header lines, we could have

`contex.setVCFHeader(vcfParser.generateVCFHeader(context.getHeaderLines));` 

Then we would generate schema by passing in the context and having it get the VCFHeader and so on.

Note that we unfortunately cannot store singular `VCFCodec` in `VcfParserImpl` (I also attempted it at first), because then it will need to be transferred to the worker, when apache beam pipeline is initiated . However, it's not serializable, so the pipeline will not be able to process it. Therefore, we are forced to generate it on the go in the `ConvertLineToVariant` class, **for each worker**. Remember, only serializable classes (with solely serializable dependencies) can be passed around into the apache beam subclasses.

Reply:
So we cannot set a global VCFCodec or VCFHeader in the pipeline runner and we are forced to store Header in every VCF record?

Reply:
No, VCFHeader (like VariantContext) is serializable, so it can be passed around. VCFCodec is not though, so it cannot.